File: js/examples/gspSignals.ts
/**
 * GSP (signaling) examples: JOIN, MUTE with CBOR args, ROLE_CHANGE.
 *
 * Demonstrates:
 *   - Signals without args via GspClient.send
 *   - Signals with per-signal CBOR args via GspClient.sendWithArgs
 *
 * Run from repo root:
 *   npx ts-node js/examples/gspSignals.ts
 */

import { MlsContext } from "../src/mls";
import { GroupNode, StreamType } from "../src/node";
import { GspClient, SignalType } from "../src/gsp";
import { PayloadCodec } from "../src/native";

// Minimal CBOR helpers.
function cborUint(n: number): number[] {
    if (n <= 23)      return [n];
    if (n <= 0xFF)    return [0x18, n];
    if (n <= 0xFFFF)  return [0x19, (n >> 8) & 0xFF, n & 0xFF];
    return [0x1A, (n>>24)&0xFF, (n>>16)&0xFF, (n>>8)&0xFF, n&0xFF];
}
function cborMap1(k: number, v: number): Buffer {
    return Buffer.from([0xA1, ...cborUint(k), ...cborUint(v)]);
}
function cborMap2(k0: number, v0: number, k1: number, v1: number): Buffer {
    return Buffer.from([0xA2, ...cborUint(k0), ...cborUint(v0), ...cborUint(k1), ...cborUint(v1)]);
}

const aliceMls = MlsContext.create("alice");
const bobMls   = MlsContext.create("bob");
bobMls.acceptWelcome(aliceMls.invite(bobMls.exportKeyPackage()));

const gid   = aliceMls.groupId;
const alice = GroupNode.create(1, gid);
const bob   = GroupNode.create(2, gid);
alice.bootstrapAsCreator(aliceMls.epoch);
bob.bootstrapAsJoiner(bobMls.epoch);

const gspAlice = GspClient.create();
const gspBob   = GspClient.create();

function recv(wire: Buffer, label: string): void {
    for (const ev of bob.onWire(bobMls, wire)) {
        if (ev.kind === "payload_received" && ev.streamType === StreamType.Signal) {
            const r = gspBob.accept(ev.plaintext!, bobMls.epoch, ev.codec ?? PayloadCodec.Cbor);
            console.log(`${label}: signal=${r.signal}  sender=${r.sender}  requestId=${r.requestId}`);
        }
    }
}

// 1. JOIN — no args.
recv(gspAlice.send(alice, aliceMls, 0, SignalType.Join, 0, 1).wire, "JOIN");

// 2. MUTE member 2 — args: {0: target_member_id=2}.
recv(gspAlice.sendWithArgs(alice, aliceMls, 0, SignalType.Mute, 0, 2, cborMap1(0, 2)).wire, "MUTE");

// 3. ROLE_CHANGE member 2 → role 1 — args: {0: target=2, 1: new_role=1}.
recv(gspAlice.sendWithArgs(alice, aliceMls, 0, SignalType.RoleChange, 1, 3, cborMap2(0, 2, 1, 1)).wire, "ROLE_CHANGE");

[aliceMls, bobMls, alice, bob, gspAlice, gspBob].forEach(x => x.close());

File: crates/gbp/stack/examples/gsp_signals.rs
//! GSP (signaling) examples: JOIN, MUTE with CBOR args, ROLE_CHANGE.
//!
//! Demonstrates:
//! - Signals without args (JOIN) via GspClient::send
//! - Signals with per-signal CBOR args (MUTE, ROLE_CHANGE) via GspClient::send_with_args
//! - Duplicate request_id is rejected with GspError::DuplicateRequest
//!
//! Run with: cargo run --example gsp_signals -p gbp-stack

use gbp_stack::{
    ControlOpcode, Event, GspClient, GroupNode, MlsContext, PayloadCodec, SignalType, StreamType,
};
use openmls::prelude::{
    DeserializeBytes as _, KeyPackageIn, OpenMlsProvider as _, ProtocolVersion,
    tls_codec::Serialize as _,
};

/// Minimal CBOR unsigned-integer encoder (no external crate needed).
fn cbor_uint(n: u32) -> Vec<u8> {
    match n {
        0..=23       => vec![n as u8],
        24..=0xFF    => vec![0x18, n as u8],
        0x100..=0xFFFF => vec![0x19, (n >> 8) as u8, n as u8],
        _ => vec![0x1A, (n >> 24) as u8, (n >> 16) as u8, (n >> 8) as u8, n as u8],
    }
}
fn cbor_map1(k: u32, v: u32) -> Vec<u8> {
    let mut b = vec![0xA1];
    b.extend(cbor_uint(k)); b.extend(cbor_uint(v)); b
}
fn cbor_map2(k0: u32, v0: u32, k1: u32, v1: u32) -> Vec<u8> {
    let mut b = vec![0xA2];
    b.extend(cbor_uint(k0)); b.extend(cbor_uint(v0));
    b.extend(cbor_uint(k1)); b.extend(cbor_uint(v1));
    b
}

fn main() -> anyhow::Result<()> {
    // --- MLS + GBP setup (identical to gtp_chat) ---------------------
    let (mut alice_mls, _) = MlsContext::new_member(b"alice")?;
    let (mut bob_mls, bob_kp_bundle) = MlsContext::new_member(b"bob")?;

    let bob_kp_bytes = bob_kp_bundle.key_package().tls_serialize_detached()?;
    let bob_kp = KeyPackageIn::tls_deserialize_exact_bytes(&bob_kp_bytes)?
        .validate(alice_mls.provider.crypto(), ProtocolVersion::Mls10)?;

    let (_commit, welcome) = alice_mls.invite_full(&[bob_kp])?;
    bob_mls.accept_welcome(&welcome)?;
    alice_mls.finalize_pending_commit()?;

    let gid = alice_mls.group_id_16();
    let mut alice = GroupNode::new(1, gid);
    let mut bob   = GroupNode::new(2, gid);
    alice.bootstrap_as_creator(0);
    bob.bootstrap_as_joiner(0, 1);

    let exec = alice.send_control(
        &mut alice_mls, 0, ControlOpcode::ExecuteTransition, 1, 7, vec![],
    )?;
    alice.apply_transition(1);
    bob.on_wire(&mut bob_mls, &exec.wire)?;

    // --- GSP clients -------------------------------------------------
    let mut gsp_alice = GspClient::new();
    let mut gsp_bob   = GspClient::new();

    // 1. JOIN — no args required.
    let frame = gsp_alice.send(
        &mut alice, &mut alice_mls,
        /*target*/ 0, SignalType::Join, /*role_claim*/ 0, /*request_id*/ 1,
        PayloadCodec::Cbor,
    )?;
    recv(&mut gsp_bob, &mut bob, &mut bob_mls, &frame.wire, "JOIN")?;

    // 2. MUTE member 2 — args: {0: target_member_id=2}.
    let frame = gsp_alice.send_with_args(
        &mut alice, &mut alice_mls,
        0, SignalType::Mute, 0, 2,
        &cbor_map1(0, 2),
        PayloadCodec::Cbor,
    )?;
    recv(&mut gsp_bob, &mut bob, &mut bob_mls, &frame.wire, "MUTE")?;

    // 3. ROLE_CHANGE member 2 → role 1 — args: {0: target=2, 1: new_role=1}.
    let frame = gsp_alice.send_with_args(
        &mut alice, &mut alice_mls,
        0, SignalType::RoleChange, /*role_claim*/ 1, 3,
        &cbor_map2(0, 2, 1, 1),
        PayloadCodec::Cbor,
    )?;
    recv(&mut gsp_bob, &mut bob, &mut bob_mls, &frame.wire, "ROLE_CHANGE")?;

    Ok(())
}

fn recv(
    gsp: &mut GspClient,
    node: &mut GroupNode,
    mls: &mut MlsContext,
    wire: &[u8],
    label: &str,
) -> anyhow::Result<()> {
    for ev in node.on_wire(mls, wire)? {
        if let Event::PayloadReceived(p) = ev {
            if p.stream_type == StreamType::Signal {
                let r = gsp.accept(&p.plaintext, mls.epoch(), p.codec)?;
                println!("{}: signal={:?}  sender={}  request_id={}",
                         label, r.signal, r.sender_id, r.request_id);
            }
        }
    }
    Ok(())
}
